ao-tools/ao-rawload/ao-rawload.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-RAWLOAD 1 "ao-rawload" ""
.SH NAME
ao-rawload \- flash a program to a AltOS device
.SH SYNOPSIS
.B "ao-rawload"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
\fIfile.ihx\fP
.SH DESCRIPTION
.I ao-rawload
loads the specified .ihx file, without modification, into the target
device flash or ram (depending on the base address of the .ihx file).
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device the debugger uses to communicate with
the target device. The special name 'BITBANG' directs ao-dbg to use
the cp2103 connection, otherwise this should be a usb serial port
connected to a suitable cc1111 debug node.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMetrum:2
.br
TeleMetrum
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.TP
\-r | --run
After the file has been loaded, set the PC to the base address of the
image and resume execution there.
the .ihx file.
.SH USAGE
.I ao-rawload
reads the specified .ihx file into memory. It then connects to the
specified target device and writes the program to the target device
memory and, optionally, starts the program executing.
.SH AUTHOR
Keith Packard


ChangeLog
commit 6e61170d42936c18cd6efba6f4c14af616a30745
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Feb 28 18:24:46 2010 -0800

    Need to duplicate new altitude conversion code in aoview.
    
    Altitude conversion is now done with a smaller table and interpolation
    instead of a giant table.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e4e17e6d3844ae682a0e7a9469a522359bac77b6
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Feb 27 17:36:13 2010 -0700

    update changelogs for Debian build

commit a1478f65538fdaac7b58ffbd958a035b74956099
Merge: 901fce5 bbf8c9f
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:19:33 2010 -0800

    Merge remote branch 'origin/master'

commit 901fce5fe3e2762406079ecaa787612b5ed4b34a
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:18:47 2010 -0800

    Add .gitignore for ao-bringup
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7aab73a265841aac817ea34235dd1eb819debf76
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:14:04 2010 -0800

    Fix up LED colors for each product.
    
    Different products assign different color LEDs to the two available
    LED drivers (P1_0, P1_1). Make the LED color pin assignments
    per-product (in ao_pins.h), then deal with not always having a green LED.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1d6d8d76ec65373ffee70add75d183f5c4168f61
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:11:44 2010 -0800

    Eliminate deadlock when writing config from radio link
    
    Writing the 'Saved\r\n' string would fill the packet buffer and cause
    a flush to occur, which would need to wait for the radio link to
    receive and transmit a packet. The radio link always re-fetches the
    radio channel number when lighting up the radio, so it need to look in
    the config space. If the config mutex was held by the config writing
    process while the radio was trying to get the channel number, then
    we'd get a deadlock.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f1956ebadd7bf46b84a05a0c383b6404ca26b344
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:10:16 2010 -0800

    Leave .ihx files in the build directory too - easier to debug that way
    
    sdcdb wants the .ihx and .cdb files in the same directory, so humor it
    by copying the .ihx files to the src directory instead of moving them.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit bbf8c9f1748af3f1fac08ddf80ae98da2e9b5727
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Feb 26 15:21:12 2010 -0700

    create a turn on script for lighting up TeleDongle v0.2 boards

commit 461d4a1948e112ec7353caf88967391d876469dd
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Feb 26 10:33:13 2010 -0800

    Add LED test

commit 5d7a3a5fbc0af4621c67a6fd51a9c9d5ae688fa5
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Feb 25 16:40:13 2010 -0800

    Fix and document the ao-rawload --run flag
    
    This allows ram-based programs to be loaded and executed easily.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1e60deca147c85a064719dfad14ccabd1049bbbd
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Feb 25 16:33:34 2010 -0800

    Allow product names to have suffixes (like board revisions)
    
    When looking for a board by product name, just look at the prefix of
    the name instead of requiring an exact match. This will allow products
    to have board version suffixes.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f4d5790a284e2d02dd7568fbca90402fa5ed1aea
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Feb 25 16:32:57 2010 -0800

    Add ao_radio_xmit to help test boards without flashing them.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit d7d551b0078acb1596a9b9023c3df6dbfa46213c
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 17:30:23 2010 -0700

    update changelogs for Debian build

commit 2f45953ee54034209a23c254e65da36e44cf075f
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 17:29:47 2010 -0700

    un-muck changelog

commit 540f86678853e7ea80ae91f3f2ec0ec88c5b1ca7
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 17:27:51 2010 -0700

    debugging

commit cdccd92f6ed97be7385e84de5694064cb8f43946
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 17:25:28 2010 -0700

    conditionalize use of git on executability of /usr/bin/git binary

commit 54dd7a200e6956aace5a580d4c4d6a10a13c654b
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 17:22:45 2010 -0700

    update changelogs for Debian build

commit 4b6b28530206c6fdbe46699b81746fbbcab5b148
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 17:22:20 2010 -0700

    see if this works

commit c74958d3e8c5ce8006a9f9f6853238fce0c77432
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 17:13:29 2010 -0700

    move git-using variable declaration into the prebuild target

commit cf091a7252eba09902a42a3cfe7b4d6a343907ce
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 16:44:44 2010 -0700

    update changelogs for Debian build

commit deccc10f0305c0cd02bf6317d6dc9d2ae38c3dac
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 16:43:32 2010 -0700

    add gawk as a build dependency since strtonum is a gawk extension

commit ac45da6e61597dcdb119b976f7301b2cc7fdced4
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 14:57:26 2010 -0700

    update changelogs for Debian build

commit 24912821b3230b8357b9e0094cd69fa0bccc5513
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Feb 24 14:53:27 2010 -0700

    update changelogs for Debian build

commit 82fdbfe1229d1ea5e6906ea240d54b3da73d9a69
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 20 21:35:43 2010 -0800

    Must install .map files for ao-load to work
    
    ao-load uses the .map files to rewrite the serial number and other

ao-tools/ao-rawload/ao-rawload.c
/*
 * Copyright © 2008 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <unistd.h>
#include <getopt.h>
#include "ccdbg.h"
#include "cc.h"

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "run", .has_arg = 0, .val = 'r' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--tty <tty-name>] [--device <device-name>] [--run] file.ihx\n", program);
	exit(1);
}

int
main (int argc, char **argv)
{
	struct ccdbg	*dbg;
	uint8_t		status;
	uint16_t	pc;
	struct hex_file	*hex;
	struct hex_image *image;
	char		*filename;
	FILE		*file;
	char		*tty = NULL;
	char		*device = NULL;
	int		c;
	int		run = 0;

	while ((c = getopt_long(argc, argv, "rT:D:", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'r':
			run = 1;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}
	filename = argv[optind];
	if (filename == NULL) {
		usage(argv[0]);
		exit(1);
	}
	file = fopen(filename, "r");
	if (!file) {
		perror(filename);
		exit(1);
	}
	hex = ccdbg_hex_file_read(file, filename);
	fclose(file);
	if (!hex)
		exit (1);
	image = ccdbg_hex_image_create(hex);
	if (!image) {
		fprintf(stderr, "image create failed\n");
		exit (1);
	}

	ccdbg_hex_file_free(hex);
	if (!tty)
		tty = cc_usbdevs_find_by_arg(device, "TIDongle");
	dbg = ccdbg_open(tty);
	if (!dbg)
		exit (1);

	ccdbg_add_debug(CC_DEBUG_FLASH);

	ccdbg_debug_mode(dbg);
	ccdbg_halt(dbg);
	if (image->address == 0xf000) {
		printf("Loading %d bytes to execute from RAM\n",
		       image->length);
		ccdbg_write_hex_image(dbg, image, 0);
	} else if (image->address == 0x0000) {
		printf("Loading %d bytes to execute from FLASH\n",
		       image->length);
		ccdbg_flash_hex_image(dbg, image);
	} else {
		printf("Cannot load code to 0x%04x\n",
		       image->address);
		ccdbg_hex_image_free(image);
		ccdbg_close(dbg);
		exit(1);
	}
	if (run) {
		ccdbg_set_pc(dbg, image->address);
		ccdbg_resume(dbg);
	}
	ccdbg_close(dbg);
	exit (0);
}
